### PR TITLE
docs/technical: add an aspirational Git submodules doc

### PR DESCRIPTION
**Pull request by chooglen**
Add `docs/technical/git-submodules.md` to capture how we think about submodules, starting with (IMO) the least controversial parts:

- Goals
- A roadmap that provides incremental value

The doc isn't complete, we will add more details and discuss them as they come, e.g. there are no designs yet.

Contributes to https://github.com/martinvonz/jj/issues/494

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: docs/technical/git-submodules.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+Colocated repositories will be supported where they are convenient, but they are
+not a high priority.
```
Comment (PhilipMetzger) [inline docs/technical/git-submodules.md:72]: I would rephrase that to something like this
```suggestion
The goal is to land good support for pure Jujutsu repositories, while Colocated repositories will be supported when convenient.
```
to deemphazise the colocated repositories to a Rust [Tier-2 or Tier-3](https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2)  target.

File: docs/technical/git-submodules.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+## Design
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
```
Comment (PhilipMetzger) [inline docs/technical/git-submodules.md:133]: Can we consider other backends before talking about the Git backend? In previous discussions it was always a hot point as we always respected Google's internal backend.

File: docs/technical/git-submodules.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: Possible, but unprioritized work
+
+- Rewriting submodule commits and updating superproject gitlinks accordingly
+
```
Comment (PhilipMetzger) [inline docs/technical/git-submodules.md:123]: Maybe add an Endgoal to summarize the ideal world? For now it can ignore all constraints, as we don't have any clues yet.

**Review by PhilipMetzger — COMMENTED** [review]
Here are some initial comments.

File: docs/technical/git-submodules.md (hunk 2 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:133]: I think I will also specify non-Git backends as a non-goal.

Though I'm not working on Google's internal backend, I feel confident saying that it will ~never need to support Git submodules.

Other, future backends are probably too ill-defined to be worth considering (e.g. Would these other backends have Git submodules? How would they convert Git submodules into their native format?). We could reconsider the feature if we start supporting non-Git 'submodules', but perhaps the general form would look very different from this proposal.

File: docs/technical/git-submodules.md [hunk 4]
```diff
@@ -65,6 +65,73 @@ pointing to `<superproject-git-directory>/modules/<submodule-name>`. The latter
...
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
```
Comment (martinvonz) [inline docs/technical/git-submodules.md:95]: This doesn't mention when the submodules are cloned or when commits are fetched. Is that intentionally left unspecified for now (fine with me)?

File: docs/technical/git-submodules.md [hunk 5]
```diff
@@ -65,6 +65,73 @@ pointing to `<superproject-git-directory>/modules/<submodule-name>`. The latter
...
+This can be done in tandem with Phase 2, but will likely require a significant
+amount of design work on its own.
+
+#### Outcomes
+
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
```
Comment (martinvonz) [inline docs/technical/git-submodules.md:122]: What do you mean by "between" here? Do you mean that a particular path is a submodule in one superproject commit and a non-submodule in another superproject commit?

**Review by martinvonz — APPROVED** [review]
Thanks!

File: docs/technical/git-submodules.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
+`<superproject-git-dir>/modules/<submodule-name>`, and a working tree with a
```
Comment (hanwen) [inline docs/technical/git-submodules.md:145]: maybe you can already establish that <name> is % encoded to avoid the bifurcation that Google's git deployment has here.

File: docs/technical/git-submodules.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+`.git` file. Using Git's own format is the simplest option, even in a
```
Comment (hanwen) [inline docs/technical/git-submodules.md:146]: it seems odd to have a `.git` file in a tree that is managed by jj.

File: docs/technical/git-submodules.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+non-colocated repository, since most (if not all) Git implementations can only
+fetch/push from a Git repository, so fetching and pushing submodules requires
+submodules to be full Git repositories.
+
+Note that submodules merely live as full Git repositories in the Git backend,
+but are not be full jj repositories. This will allow us to create a global view
+over the superproject and submodules, which should allow us to avoid some of
+Git's pitfalls (e.g. there is no obvious way to tell the submodule to preserve
+commits because they are needed by the superproject). Compared to Git, this
```
Comment (hanwen) [inline docs/technical/git-submodules.md:155]: is this is a concern for JJ ? JJ already has refs/jj/keep to keep things alive. 

I have been mulling a bit over this too, and actually thought it would be easier to have a `.jj/modules/$NAME`, which contains a JJ repository. Since Git isn't run directly on the git repo within the jj repo, is there a benefit to keeping the submodules in the git repo?

By contrast, if you use the recursive solution, it means you can get JJ to work on the inner repos for free.

File: docs/technical/git-submodules.md (hunk 7 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:146]: it also creates a risk that people can run `git` in the jj submodule and create havoc by modifying things outside of JJ's knowledge.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:155]: > is this is a concern for JJ ? JJ already has refs/jj/keep to keep things alive.

Yeah, gc probably won't be a problem for jj submodules for this reason.

> I have been mulling a bit over this too, and actually thought it would be easier to have a .jj/modules/$NAME, which contains a JJ repository.

I think 'global view' of the repo is valuable for other reasons mostly. @martinvonz brought up the case of the operation log/`jj undo`. Having a single operation log track changes in the submodule and superproject would probably simplify how we implement this vs coordinating between the operation logs of the jj superproject and jj submodule.

I expect a trickle of unknown unknown reasons why we need to coordinate between the superproject and submodule, so I'd still prefer to have the superproject as the entrypoint for everything. That said, maybe it's not so hard to have full jj repos and coordinate between them. I'm not so sure either way, WDYT?

> Since Git isn't run directly on the git repo within the jj repo, is there a benefit to keeping the submodules in the git repo?

Even in a jj-only repository, we do run some Git-like operations in the Git repo (e.g. fetch, push). In _theory_ these could have some smart integrations with submodules, which we might want for `jj git [fetch|push]`. In practice, I don't think these integrations exist (not for `git/git` at least), and I don't know what libgit2 (or the Rust bindings) offer us in this regard (probably not much).

> By contrast, if you use the recursive solution, it means you can get JJ to work on the inner repos for free.

True

File: docs/technical/git-submodules.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+The goal is to land good support for pure Jujutsu repositories, while colocated
+repositories will be supported when convenient.
```
Comment (chooglen) [inline docs/technical/git-submodules.md:76]: I realize that this statement on colocation might not capture how I actually think about the roadmap. I initially wrote this thinking that the pure jj case and colocated case wouldn't be that different, so this was more of an escape hatch in case something about colocation turned out to be more difficult than expected.

But I'm not sure if this 'incremental use case' sort of roadmap makes a lot of sense without colocated repo support. I'd expect Git submodule early adopters to use colocated repos since they can fall back to Git for the use cases jj doesn't support. On the other hand, if we only handle pure jj repositories, we'd need an almost complete implementation before users can actually do anything with submodules.

What do others think? Does this roadmap make sense even if we only support jj-only?

Comment (martinvonz) [reply docs/technical/git-submodules.md:76]: I would hope that simply populating the submodules in the working copy would be enough for many users, but I don't use any repos with submodules, so I don't really know. Is it common that users have to modify submodules (not just the superproject)?

Comment (PhilipMetzger) [reply docs/technical/git-submodules.md:76]: > I realize that this statement on colocation might not capture how I actually think about the roadmap. I initially wrote this thinking that the pure jj case and colocated case wouldn't be that different, so this was more of an escape hatch in case something about colocation turned out to be more difficult than expected.

I think adding the escape hatch is a good idea, it gives us an out when we're unable to progress. But your point stands that we'd get no direct feedback, when decisions heavily depend on it. So I'd say we "officially" mention the pure use-case as the final goal, while keeping colocated repositories as best effort. This allows a clean design, while getting the incremental feedback you wanted.

But I've never used submodules.

File: docs/technical/git-submodules.md (hunk 2 above)
Comment (PhilipMetzger) [reply docs/technical/git-submodules.md:144]: > I think I will also specify non-Git backends as a non-goal.

Thats fair but I'd still like to discuss the impact of your changes to the base backend. This is to avoid leaking internals to other backends and building a common understanding, of what a backend is/implies. The goal here is to take a step back and minimize the impact on the base backend.

> Though I'm not working on Google's internal backend, I feel confident saying that it will ~never need to support Git submodules.

I also don't expect something Perforce-like to ever grow a understanding of Git submodules. But Git submodules cannot constrain it, or any other backend. 

> Other, future backends are probably too ill-defined to be worth considering (e.g. Would these other backends have Git submodules? How would they convert Git submodules into their native format?). We could reconsider the feature if we start supporting non-Git 'submodules', but perhaps the general form would look very different from this proposal.

I don't expect that any other backend will implement something like Git submodules or Jujutsu submodules if they ever exist. So it's not immediately worth discussing, but we'd still learn something while doing some space exploration. But its totally out of scope for this proposal.

Comment (chooglen) [reply docs/technical/git-submodules.md:144]: > Thats fair but I'd still like to discuss the impact of your changes to the base backend. This is to avoid leaking internals to other backends and building a common understanding, of what a backend is/implies. The goal here is to take a step back and minimize the impact on the base backend.

Yes, I've considered that as well, though I'm not concerned since there are design patterns that can manage this. A successful one IMO is "capabilities", where we could specify a set of capabilities that a backend may or may not support, and Git submodules can be one of them. Other backends that want to support Git submodules can provide their own implementation, but it won't constrain backends that don't support submodules.

But this does raise the question of whether the capability should be "Git submodules", or a generalized form ("Subrepos"?). I'm leaning towards "Git submodules", because I think it would be difficult to generalize "subrepos" across all possible backends.

That said, there is an alternative in this space..

> I don't expect that any other backend will implement something like Git submodules or Jujutsu submodules if they ever exist. So it's not immediately worth discussing, but we'd still learn something while doing some space exploration. But its totally out of scope for this proposal.

Which is to implement Git submodules as jj repositories, as @hanwen suggested https://github.com/martinvonz/jj/pull/1411#discussion_r1144994721. I think that already gets us to generalized "subrepos", because we can tell the subrepo what to do and not worry about its backend.

My biggest concern with that proposal is that the layout on disk would probably be different from what Git expects, so it might make colocated repos harder to implement. _But_ I feel fairly confident that we can upstream changes Git and libgit2 to understand the different layout, so I'm not that worried.

Comment (PhilipMetzger) [reply docs/technical/git-submodules.md:144]: > Yes, I've considered that as well, though I'm not concerned since there are design patterns that can manage this. A successful one IMO is "capabilities", where we could specify a set of capabilities that a backend may or may not support, and Git submodules can be one of them. Other backends that want to support Git submodules can provide their own implementation, but it won't constrain backends that don't support submodules.

I'm not directly concerned with the trait or interface but rather with the concept. I agree that there are many design patterns which help to solve this issue.

> But this does raise the question of whether the capability should be "Git submodules", or a generalized form ("Subrepos"?). I'm leaning towards "Git submodules", because I think it would be difficult to generalize "subrepos" across all possible backends.
[...]
> Which is to implement Git submodules as jj repositories, as @hanwen suggested https://github.com/martinvonz/jj/pull/1411#discussion_r1144994721. I think that already gets us to generalized "subrepos", because we can tell the subrepo what to do and not worry about its backend.

This sounds like something for `Alternatives Considered` section, if we were discussing a JJ submodules proposal, its also very neat. 

> My biggest concern with that proposal is that the layout on disk would probably be different from what Git expects, so it might make colocated repos harder to implement. But I feel fairly confident that we can upstream changes Git and libgit2 to understand the different layout, so I'm not that worried.

I'm also not worried about a disk format yet, as this is something for the detailed implementation plan. I assume your capable of doing the necessary changes in Git and JJ, since you're a Git contributor.

File: docs/technical/git-submodules.md (hunk 6 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:145]: I'm sorry, I was confused about the general gist of your idea. Git's tendency to put things called `$NAME` into `.git/somedir/$NAME` is a bad idea for the following reasons:

* security (what if $NAME = "../../blah")
* portability (what about case-insensitive vs case-sensitive filesystems?)
* functionality (dir/file conflicts)

I was hoping that JJ's approach wouldn't make the same mistake here, and could specify something that always escapes the name.

File: docs/technical/git-submodules.md (hunk 9 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:76]: > so I don't really know. Is it common that users have to modify submodules (not just the superproject)?

the major use-case (inside Google) would be Android development. The superproject (https://googleplex-android.git.corp.google.com/platform/superproject) is just a config that stitches together ~1000 submodules. Most work on android touches just a single submodule, but about 10% needs changes that coordinate across submodules.

Other examples that I know of are Gerrit. There we have a submodule for JGit (a base dependency), and for plugins,  (https://gerrit.googlesource.com/gerrit/+/refs/heads/master/plugins/).

JGit development typically happens in the JGit repo, but it's useful for prototyping/debugging to make changes in the submodule too.

For plugins, code refactoring in Gerrit main repo (which provides APIs) may cause changes in the plugins (consumers of the APIs), and they need to be coordinated.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:155]: > Even in a jj-only repository, we do run some Git-like operations in the Git repo (e.g. fetch, push). In theory these could have some smart integrations with submodules, which we might want for jj git [fetch|push]. In practice, I don't think these integrations exist (not for git/git at least), and I don't know what libgit2 (or the Rust bindings) offer us in this regard (probably not much).

My experience is that you don't want your base library (libgit2 in this case) to have 'smart' features. What git/git thinks is a sensible default will probably get in the way in a project that uses git in a different way. If JJ knows what submodules there are, it can coordinate operations across the submodules. The storage layer/API doesn't permit cross-repo operations, so I don't think git/git can essentially do better.

File: docs/technical/git-submodules.md (hunk 7 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:146]: > I would also prefer not to have two different implementations for the jj-only and colocated cases. WDYT?

what is the long-term target? colocated sounds like it is an interim solution to ease the transition to a jj-only world?

**Comment by PhilipMetzger** [comment]
Can we move the document from `docs/technical` to `docs/design` or similar? Then we'll have a place for other design docs, which exist but aren't implemented yet.

File: docs/technical/git-submodules.md (hunk 5 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:122]: Yes. Is there a better way to express that?

File: docs/technical/git-submodules.md (hunk 4 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:95]: > This doesn't mention when the submodules are cloned or when commits are fetched. Is that intentionally left unspecified for now (fine with me)?

Yes, it is intentionally underspecified. I have some thoughts on how we might do this more ergonomically than Git, but they deserve their own discussion, so I'll send those thoughts as a separate PR.

File: docs/technical/git-submodules.md (hunk 6 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:145]: Ah I see, I misunderstood. You are suggesting that `jj` should preemptively %-encode the name, using Google's Git patches as a example of what _not_ to do. I think this makes sense. Doing this means that in colocated repos, Git won't understand the submodule directory name, but we can upstream the change into Git. (That change will probably be welcome by the rest of the Google Git team, since we can drop the downstream patch.)

File: docs/technical/git-submodules.md (hunk 7 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:146]: My long-term target is to support both colocated and jj-only repos.

> colocated sounds like it is an interim solution to ease the transition to a jj-only world?

I suppose that is true, but I am not so confident that the world will converge onto jj (as much as I would like that), or at least, not quickly enough. I expect that having a Git repo as a common denominator will be useful for a very very long time.

File: docs/technical/git-submodules.md (hunk 2 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:144]: > I'm not directly concerned with the trait or interface but rather with the concept.

Hm, I suppose I'm having trouble imagining what you're describing here. Could you perhaps give an example sentence or two of what you wish I had included?

At any rate, I'm thinking of taking a step back from the idea of having Git submodules in the Git backend only. I suggested it not because it follows from the primary requirements, but because I thought it was uncontroversial for submodules to not be jj repos (based on informal discussions), so this would be the easiest way to get to colocated support . However, @hanwen has given some good reasons to think otherwise, so I'll probably revise this.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:155]: > Isn't it too early to make a call here? Ideally, the decision follows from requirements.

Yes, I think you're right. Based on informal discussions, I thought that storing submodules in the Git backend would be uncontroversial, but you've provided good reasons to question that, so I'll step back from this proposal for now.

> it would be great if we could somehow do some experimentation to see what workflow works well in practice, before committing to a decision.

Yes, I'm trying to think of what experiment would be the most educational; I'll probably send another round of RFC to get ideas. Perhaps something like `jj op log` across superproject and submodules would be the most informative, but we'd have to invest substantially into submodules before we could meaningfully test that.

Alternatively, perhaps the better approach is just whichever is less invasive, so that it's easier to undo in case we decide the other approach is better.

File: docs/technical/git-submodules.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
+# Git submodules
+
+This is an aspirational document that describes how jj _will_ support Git
+submodules. Readers are assumed to have some familiarity with Git and Git
+submodules.
+
+This document is a work in progress; submodules are a big feature, and relevant
```
[the remark this answers is not in the document]
Comment (chooglen) [reply docs/technical/git-submodules.md:7]: Uh... I don't know. `submodules are` reads more correctly to me..?

If multiple submodules comprise the feature, then I think referring to them collectively with `are` is correct.
But if we are talking about a singular feature, then `is` is probably correct.

I don't which is more applicable in this case.

File: docs/technical/git-submodules.md (hunk 9 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:76]: In the Gerrit scenario, would it be valuable to have readonly submodules in jj? I'd imagine it's possible that changes in the submodule don't happen that frequently, so there might be users who would use jj if it allowed them to just build the submodule.

File: docs/technical/git-submodules.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: An ideal world
+
+I.e. outcomes we would like to see if there were no constraints whatsoever.
+
+- Rewriting submodule commits rewrites descendants correctly and updates
+  superproject gitlinks.
```
Comment (martinvonz) [inline docs/technical/git-submodules.md:129]: Interesting. When modifying a regular file in a repo without submodules, that modification is always in the context of a commit, and other branches with the same version of the file will not be rewritten. I suppose we'll want to do the same for submodules. I suspect you'll naturally end up implementing it that way by just following the patterns for how parent trees are updated with changes from children.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (martinvonz) [reply docs/technical/git-submodules.md:155]: I'm also unsure if the top-level repo's operation log should record the information about submodules' view objects (refs etc.) or if that should be tracked separately for each submodule and the top-level should just track the operation ids (or view ids?) of the submodules. It might be easiest to implement it as full recursive repos (as @hanwen suggested). We might still want to ignore what the submodules' think is the latest operation according to their "op heads store", though. The superproject should probably be considered the source of truth for that.

File: docs/technical/git-submodules.md (hunk 2 above)
Comment (PhilipMetzger) [reply docs/technical/git-submodules.md:144]: Sorry for being so obtuse, I was a bit caught in my own thoughts. 

>> I'm not directly concerned with the trait or interface but rather with the concept.

>Hm, I suppose I'm having trouble imagining what you're describing here. Could you perhaps give an example sentence or two of what you wish I had included?

In my opinion, you went straight to a implementation with your idea (which is fine), but you brushed my concerns about a larger impact on Jujutsu aside. This by moving it out of scope by making it Git only[^1] or capabilites[^2].  

I just wanted to discuss the benefits or drawbacks for Jujutsu and other backends, without concerns about how an implementation would look 

> At any rate, I'm thinking of taking a step back from the idea of having Git submodules in the Git backend only. I suggested it not because it follows from the primary requirements, but because I thought it was uncontroversial for submodules to not be jj repos (based on informal discussions), so this would be the easiest way to get to colocated support . However, @hanwen has given some good reasons to think otherwise, so I'll probably revise this.  

I also think @hanwen found a good solution and I'll be happy to discuss it further after an Google internal discussion.

> so, in general, your doc starts off with a set of requirements, which I think is good, but in this section you're already pinning yourself to a technical decision which will have a lot of future implications. Isn't it too early to make a call here? Ideally, the decision follows from requirements.

This also reflects my opinion. 

Sorry for the Wall of Text, I just wanted to leave no room for misinterpretations. 


[^1]: Aside, then the proposal should be called `Git submodules for the Git backend`, as the goal is to improve the UX.
[^2]: Which address the problem on an interface/trait level, sidestepping the discussion on potential global impact.

File: docs/technical/git-submodules.md (hunk 9 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:76]: having readonly submodules would allow the Gerrit team dogfood JJ. Changes to plugins or JGit are uncommon enough.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (hanwen) [reply docs/technical/git-submodules.md:155]: FWIW, I've been toying with a FUSE filesystem with embedded version control command (inspired by JJ), which implements the full recursive solution.  I haven't gotten round to implementing merge, but might be useful to test some of these options. It has no users, so there is no compatibility to worry about :-)

File: docs/technical/git-submodules.md (hunk 9 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:76]: Perfect. It sounds like this roadmap still makes sense in a jj-only world. Thanks!

File: docs/technical/git-submodules.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
```
[the remark this answers is not in the document]
Comment (chooglen) [reply docs/technical/git-submodules.md:95]: Good question. I don't really like the inconsistency either; I use it out of habit, and the Git docs say "superproject" if they say anything at all. My preference would be to stick with Git terminology for the Git-specific thing so that we'll leave the nicer terminology alone in case we ever implement generic jj subrepos.

File: docs/technical/git-submodules.md (hunk 8 above)
Comment (chooglen) [reply docs/technical/git-submodules.md:155]: > We might still want to ignore what the submodules' think is the latest operation according to their "op heads store", though. The superproject should probably be considered the source of truth for that.

That's quite promising. I think that would address a major problem with Git + submodules (that I was also trying to eliminate with the 'single operation log' idea), which is that there's no way to transactionally do anything across the superproject and submodules (e.g. if the command succeeds in sub1 but fails in sub2, there isn't a good way to roll back sub1) because each repo is independent and everything is just done via subprocesses. If we take your suggestion where the superproject's operation log tracks the submodule's operation, the superproject can just roll back failed commands by setting the submodule's latest operation.

File: docs/technical/git-submodules.md (hunk 2 above)
Comment (PhilipMetzger) [reply docs/technical/git-submodules.md:144]: > I'm sorry if I gave that impression. That was absolutely not my intention; it's quite clear now that I misinterpreted you quite a lot.

No harm done :-), we just talked past each other.

 > I'm afraid I still don't completely follow :/ I read this comment as "the benefit or drawbacks for Jujutsu and other backends (of supporting Git submodules), but rereading your initial comment, my impression is that you were talking about what considering support [Git] submodules in other backends.

You got the point, which is important. Even writing two to three sentences about other backends and dismissing them would've been enough, to bring out concerns which weren't considered before. 

> Tbh I'd prefer for the discussion to not be internal-only. After all, we aren't the only ones using jj :)

I just deferred to the internal discussion as you have some Git submodule users already (Android, Gerrit and Chromium in a year or two). Its helpful for everyone if you'd evaluate some proposed solutions against your use cases. 
As I'm neither user nor have any direct use case for Git submodules.   

> If there are no objections, I'll scale this PR back to the truly uncontroversial bits (which seem to be requirements + roadmap). Then, I'll focus my next PR on figuring out the criteria that we'll use to decide which solution to go with, and evaluating possible approaches.

I do not have any further objections. 

Although an experience report of your Submodule work in git/git could be useful in the next PR, but that's totally up to you.

Comment (chooglen) [reply docs/technical/git-submodules.md]: > Even writing two to three sentences about other backends and dismissing them would've been enough, to bring out concerns which weren't considered before.

Ah, yes, that makes sense. I was counting on the shared context that others + myself built up over informal conversations, but now I see how jarring this is if you're not privy to that context.

> I do not have any further objections.

Thanks. Let's move this discussion to the follow up PR :)
